README.md

```diff
@@ -191,6 +191,65 @@ input.dispatchEvent(
 );
 ```
 
+## Adding TypeScript typed events
+
+If you're using TypeScript, you can maintain a list of custom event names that map to their specific types, making it easier to write type-safe code while using delegated events. Add the following snippet to a `.d.ts` file in your local project and alter the contents of `CustomEventMap` to list all the well-known events in your project:
+
+```typescript
+// events.d.ts
+interface CustomEventMap {
+  'my-event:foo': {
+    something: boolean
+  }
+  // When adding a new custom event to your code, add the event.name + event.detail type to this map!
+}
+
+// Do not change code below this line!
+type CustomDelegatedEventListener<T> = (this: Element, ev: CustomEvent<T> & {currentTarget: Element}) => any
+
+declare module 'delegated-events' {
+  export function fire<K extends keyof CustomEventMap>(target: Element, name: K, detail: CustomEventMap[K]): boolean
+  export function on<K extends keyof CustomEventMap>(
+    name: K,
+    selector: string,
+    listener: CustomDelegatedEventListener<CustomEventMap[K]>
+  ): boolean
+}
+
+declare global {
+  interface Document {
+    addEventListener<K extends keyof CustomEventMap>(
+      type: K,
+      listener: (this: Document, ev: CustomEvent<CustomEventMap[K]>) => unknown,
+      options?: boolean | AddEventListenerOptions
+    ): void
+  }
+  interface HTMLElement {
+    addEventListener<K extends keyof CustomEventMap>(
+      type: K,
+      listener: (this: HTMLElement, ev: CustomEvent<CustomEventMap[K]>) => unknown,
+      options?: boolean | AddEventListenerOptions
+    ): void
+  }
+}
+```
+
+Now TypeScript is able to type-check your events in both `delegated-events` callsites and the standard `addEventListener` callsites:
+
+```typescript
+fire(document.body, 'my-event:foo', {something: true})
```

Comment:
I wasn't having any luck with type checking or autocomplete in vscode unless I included the generic constraint. Not sure if that's specific to my setup or not though.

Reply:
The generic constraint that comes from the typescript definition within this file? Interesting...

Reply:
If I leave the generic constraint off I get autocomplete for the event name, but not on the details. I also don't get type checking for the details. If I adjust it like so then I get auto complete for the event name & details, along with type checking for the details.

```typescript
fire<'my-event:foo'>(document.body, 'my-event:foo', {something: true})
```

The rest of the functions works as expected, it's just the `fire()` one that's not.

I'm using v3.9.6 & v3.9.7 of TypeScript. I'll try rolling back a few versions to see if there might have been a regression at some point.